src/ethereum_spec_tools/evm_tools/t8n/t8n_types.py

```diff
@@ -0,0 +1,348 @@
+        # Store a maximum of 256 block hashes.
+        max_blockhash_count = min(256, self.block_number)
+        for number in range(
+            self.block_number - max_blockhash_count, self.block_number
+        ):
+            if "blockHashes" in data and str(number) in data["blockHashes"]:
+                block_hashes.append(
+                    Hash32(hex_to_bytes(data["blockHashes"][str(number)]))
+                )
+            else:
+                block_hashes.append(None)
+
+        self.block_hashes = block_hashes
+
+    def read_ommers(self, data: Any, t8n: Any) -> None:
+        """
+        Read the ommers. The ommers data might not have all the details
+        needed to obtain the Header.
+        """
+        ommers = []
+        if "ommers" in data:
+            for ommer in data["ommers"]:
+                ommers.append(
+                    Ommer(
+                        ommer["delta"],
+                        t8n.hex_to_address(ommer["address"]),
+                    )
+                )
+        self.ommers = ommers
+
+
+class Alloc:
+    """The alloc (state) type for the t8n tool."""
+
+    state: Any
+    state_backup: Any = None
+
+    def __init__(self, t8n: Any):
+        """Read the alloc file and return the state."""
+        with open(t8n.options.input_alloc, "r") as f:
+            data = json.load(f)
+
+        # The json_to_state functions expects the values to hex
+        # strings, so we convert them here.
+        for address, account in data.items():
+            for key, value in account.items():
+                if key == "storage":
+                    continue
+                elif not value.startswith("0x"):
+                    data[address][key] = "0x" + hex(int(value))
+
+        self.state = t8n.json_to_state(data)
+
+    def to_json(self) -> Any:
+        """Encode the state to JSON"""
+        data = {}
+        for address, account in self.state._main_trie._data.items():
+
+            account_data: Dict[str, Any] = {}
+
+            if account.balance:
+                account_data["balance"] = hex(account.balance)
+
+            if account.nonce:
+                account_data["nonce"] = hex(account.nonce)
+
+            if account.code:
+                account_data["code"] = account.code.hex()
+
+            if address in self.state._storage_tries:
+                account_data["storage"] = {
+                    k.hex(): hex(v)
+                    for k, v in self.state._storage_tries[
+                        address
+                    ]._data.items()
+                }
+
+            data["0x" + address.hex()] = account_data
+
+        return data
+
+
+class Txs:
+    """
+    Read the transactions file, sort out the valid transactions and
+    return a list of transactions.
+    """
+
+    rejected_txs: Any = None
+    t8n: Any = None
+
+    def __init__(self, t8n: Any):
+        self.t8n = t8n
+        self.rejected_txs = {}
+
+    @property
```

Comment:
Properties should generally be used for returning data quickly. For example, it wouldn't be unreasonable for a dev to do:

```python
count = len(foo.transactions)
for tx in foo.transactions:
    ...
```

Which would cause the file to be read twice.

Reply:
Nice catch. Have moved the file read into the `__init__` function to avoid multiple reads.